File: test/org/cerion/symcalc/expression/function/list/PartitionTest.java
package org.cerion.symcalc.expression.function.list;

import org.cerion.symcalc.expression.Expr;
import org.cerion.symcalc.expression.ListExpr;
import org.cerion.symcalc.expression.number.IntegerNum;
import org.junit.Test;

import static org.junit.Assert.*;

public class PartitionTest {

    @Test
    public void validate() {

    }

    @Test
    public void basicTest() {
        // Same length
        ListExpr list = new ListExpr(IntegerNum.ZERO, IntegerNum.ONE, IntegerNum.TWO, new IntegerNum(3));
        Expr e = new Partition(list, new IntegerNum(4)).eval();
        assertEquals(new ListExpr(list), e);

        // Split in half
        e = new Partition(list, IntegerNum.TWO).eval();

        ListExpr expected = new ListExpr(
                new ListExpr(IntegerNum.ZERO, IntegerNum.ONE),
                new ListExpr(IntegerNum.TWO, new IntegerNum(3)));

        assertEquals(expected, e);
    }

    @Test
    public void listTooShort() {
        // Simpler version of next test where the first sublist is too short

        ListExpr list = new ListExpr(IntegerNum.ZERO, IntegerNum.ONE, IntegerNum.TWO);
        Expr e = new Partition(list, new IntegerNum(4)).eval();

        ListExpr expected = new ListExpr();
        assertEquals(expected, e);
    }

    @Test
    public void truncateExtraItems() {
        ListExpr list = new ListExpr(IntegerNum.ZERO, IntegerNum.ONE, IntegerNum.TWO, new IntegerNum(3), new IntegerNum(4));
        Expr e = new Partition(list, new IntegerNum(2)).eval();

        ListExpr expected = new ListExpr(
                new ListExpr(IntegerNum.ZERO, IntegerNum.ONE),
                new ListExpr(IntegerNum.TWO, new IntegerNum(3)));

        assertEquals(expected, e);
    }
}

File: test/expression/function/list/PartitionTest.kt
package org.cerion.symcalc.expression.function.list

import org.cerion.symcalc.expression.ListExpr
import org.cerion.symcalc.expression.number.Integer
import org.junit.Assert.assertEquals
import org.junit.Test

class PartitionTest {

    @Test
    fun validate() {

    }

    @Test
    fun basicTest() {
        // Same length
        val list = ListExpr(Integer.ZERO, Integer.ONE, Integer.TWO, Integer(3))
        var e = Partition(list, Integer(4)).eval()
        assertEquals(ListExpr(list), e)

        // Split in half
        e = Partition(list, Integer.TWO).eval()

        val expected = ListExpr(
                ListExpr(Integer.ZERO, Integer.ONE),
                ListExpr(Integer.TWO, Integer(3)))

        assertEquals(expected, e)
    }

    @Test
    fun listTooShort() {
        // Simpler version of next test where the first sublist is too short

        val list = ListExpr(Integer.ZERO, Integer.ONE, Integer.TWO)
        val e = Partition(list, Integer(4)).eval()

        val expected = ListExpr()
        assertEquals(expected, e)
    }

    @Test
    fun truncateExtraItems() {
        val list = ListExpr(Integer.ZERO, Integer.ONE, Integer.TWO, Integer(3), Integer(4))
        val e = Partition(list, Integer(2)).eval()

        val expected = ListExpr(
                ListExpr(Integer.ZERO, Integer.ONE),
                ListExpr(Integer.TWO, Integer(3)))

        assertEquals(expected, e)
    }
}